src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -138,18 +169,36 @@ func (a *StorageMinerActorCode_I) _submitPowerReport(rt Runtime, lastPoSt block.
 	activePower := st._getActivePower(rt)
 	inactivePower := st._getInactivePower(rt)
 
-	powerReport := &power.PowerReport_I{
+	// serialize this in param
+	_ = &power.PowerReport_I{
 		ActivePower_:   activePower,
 		InactivePower_: inactivePower,
 	}
 
+	// TODO: serialization helper
+	processPowerReportParam := make([]actor.MethodParam, 1)
+	processDealExpirationParam := make([]actor.MethodParam, len(newExpiredDealIDs))
+
 	Release(rt, h, st)
 
-	rt.Abort("TODO") // TODO: Send(SPA, ProcessPowerReport(powerReport))
-	panic(powerReport)
+	// this will go through even if miners do not have the right amount of pledge collateral
+	// when _submitPowerReport is called in DeclareFaults and _onMissedSurprisePoSt for power slashing
+	// however in _onSuccessfulPoSt EnsurePledgeCollateralSatsified will be called
+	// to ensure that miners have the required pledge collateral
+	// otherwise, post submission will fail
+	// Note: there is no power update in RecoverFaults and hence no EnsurePledgeCollatera or _submitPowerReport
+	rt.SendCatchingErrors(&msg.InvocInput_I{
+		To_:     addr.StoragePowerActorAddr,
+		Method_: power.MethodProcessPowerReport,
+		Params_: processPowerReportParam,
+	})
 
 	if len(newExpiredDealIDs) > 0 {
-		// Send(StorageMarketActor, ProcessDealExpiration)
+		rt.SendCatchingErrors(&msg.InvocInput_I{
+			To_:     addr.StorageMarketActorAddr,
+			Method_: market.MethodProcessDealExpiration,
+			Params_: processDealExpirationParam,
+		})
```

Comment:
Ditto for both of the above: we have to write code that works, not defensive checks around code that doesn't. These methods should not fail for any reason preventable by the programmer, and unpreventable errors should revert the entire state change.

Reply:
+1. Also want to flag that we should carefully analyze what happens upon failure (even out-of-gas failures, which we can't prevent in general), given that these are messages ultimately originated by users. I.e., the consequences of any type of failure should be no worse than the result if the user had just not submitted the message in the first place.